4055 add contacts mobile

Faldrian opened the pull request:
Added a "add contact" dropdown in mobile view.
A user can add and remove people from existing aspects in mobile view.
I use the first option in the select field to be a cover reflecting the current state (like the desktop-dropdown does). Since this option is marked disabled and shown in the popup, the user has some overview of the current status when adding a person to aspects.

**Why select in single-select-mode and not "multiple"?**
Adding people to aspects is a "choose multiple" task, similar to selecting aspects in the publisher. The multiple-select HTML-element would be perfect for that. Instead I did some JS-magic with the single-mode-select because mobile browsers are much more predictable how they display the select (firefox mobile was not very usable with multiple select).

**The checks and dots are not well aligned!**
Yeah... since I did not find a way to style the selection popup. This is the best I could do with utf8 to let it look okay... If someone has a better idea, please comment.

This will solve #4055.

(I have to write tests now...)

Screenshots (Firefox mobile on Android):
![screenshot_2015-01-27-10-35-16](https://cloud.githubusercontent.com/assets/1184859/5916331/5414924a-a610-11e4-8641-9fe10196bcdf.png)

![profile2](https://cloud.githubusercontent.com/assets/1184859/5902945/03e432d0-a57f-11e4-87b4-792630cf40da.png)

Google Chrome on Android:
![screenshot_2015-01-27-10-38-05](https://cloud.githubusercontent.com/assets/1184859/5916372/b6ca56c2-a610-11e4-855f-a333fb63b803.png)

Faldrian committed:
Added dropdown to add/remove people from/to aspects in mobile view

Flaburgan commented:
You're the man!

Flaburgan commented:
Would you mind indicating [on bountysource](https://www.bountysource.com/issues/192899-unable-to-add-contacts-in-mobile) that you are working on this? You will help us know how this tool works :p

SansPseudoFix commented:
Yeah, big thanks for that ! Many users want that feature !

Faldrian commented:
@SansPseudoFix I know, that's why it was now top of my list and I got time to do it. :)

Faldrian commented:
Added tests, edited changelog and squashed into one commit.

Faldrian commented:
The Hovercard-Test again... hmpf. Needs to be more robust.

Flaburgan commented:
Hey, I tried to test that locally, so switching to the mobile version on Firefox desktop, and when I click on an aspect I got a popup with "Failed to add contact to aspect."

Here are the logs:

<pre>
Started POST "/aspect_memberships" for 127.0.0.1 at 2015-01-27 09:55:57 +0100
Processing by AspectMembershipsController#create as JSON
  Parameters: {"person_id"=>2, "aspect_id"=>6, "aspect_membership"=>{"aspect_id"=>6}}
  ESC[1mESC[36mUser Load (0.8ms)ESC[0m  ESC[1mSELECT  "users".* FROM "users"  WHERE "users"."id" = 2  ORDER BY "users"."id" ASC LIMIT 1ESC[0m
  ESC[1mESC[35mPerson Load (0.4ms)ESC[0m  SELECT  "people".* FROM "people"  WHERE "people"."id" = $1 LIMIT 1  [["id", 2]]
  ESC[1mESC[36mAspect Load (0.5ms)ESC[0m  ESC[1mSELECT  "aspects".* FROM "aspects"  WHERE "aspects"."user_id" = $1 AND "aspects"."id" = 6  ORDER BY order_id ASC LIMIT 1ESC[0m  [["user_id", 2]]
  ESC[1mESC[35mContact Load (0.5ms)ESC[0m  SELECT  "contacts".* FROM "contacts"  WHERE "contacts"."user_id" = $1 AND "contacts"."person_id" = 2 LIMIT 1  [["user_id", 2]]
  ESC[1mESC[36mPerson Load (0.3ms)ESC[0m  ESC[1mSELECT  "people".* FROM "people"  WHERE "people"."id" = $1 LIMIT 1ESC[0m  [["id", 2]]
  ESC[1mESC[35mUser Load (0.5ms)ESC[0m  SELECT  "users".* FROM "users"  WHERE "users"."id" = $1 LIMIT 1  [["id", 2]]
  ESC[1mESC[36mBlock Exists (0.4ms)ESC[0m  ESC[1mSELECT  1 AS one FROM "blocks"  WHERE "blocks"."user_id" = $1 AND "blocks"."person_id" = 2 LIMIT 1ESC[0m  [["user_id", 2]]
  ESC[1mESC[35mContact Exists (0.4ms)ESC[0m  SELECT  1 AS one FROM "contacts"  WHERE ("contacts"."person_id" = 2 AND "contacts"."user_id" = 2) LIMIT 1
  Rendered text template (0.0ms)
Completed 409 Conflict in 17ms (Views: 0.5ms | ActiveRecord: 3.8ms)
</pre>

Also, I think it would be nice to start breaking the mobile.js file in different files.

Faldrian commented:
I will check that.

Breaking mobile.js in different files is another issue, you can open an issue for that. I also think we have to refactor some things for mobile.js, as there is no i18n support for mobile.js so far (and no backbone is used.... which is okay, because mobile should be a light interface, but also lacks the benefit of using all desktop-version goodies).

Flaburgan commented:
>Breaking mobile.js in different files is another issue, you can open an issue for that.

Well, I do not ask you to break the existing file in that PR, just to put your code in "aspect_management.js" (or a better name) instead of mobile.js ;)

Faldrian commented:
@Flaburgan I moved it to `profile_aspects.js` (since we don't manage aspects, we have things related to aspects on the profile page... so I thought this name would be suitable).

Also thank you VERY much for your local test ... I am a bit ashamed, but I left some debugging hardcoded person-id in the code and it worked for me because my tests seems to use the same ids as my test accounts.... fixed that now.

Flaburgan commented:
Confirmed, it now works. Thank you!

It would be nice if you can add some visual improvement to the aspect drop down to match the desktop view, especially to have it green when the contact is already in at least one aspect, an gray if he isn't.

Faldrian commented:
@Flaburgan I tested that yesterday... but I tried it again and now the dropdown will be green if the person is in aspects. There is no gradient, because select-fields seem not to support gradients as background so far. :)

Flaburgan commented:
No problem about the gradient. To add the caret would be nice though. I would also remove the float: right;

Faldrian commented:
caret?

float: right is necessary, so the button appears on the right like in desktop view.

Flaburgan commented:
Here is a caret: &#9660; it's `&#9660;`

I personally find 
![capture du 2015-01-27 10 42 46](https://cloud.githubusercontent.com/assets/930064/5916429/7da9a5ae-a611-11e4-97bd-dee5fa31dd58.png)


nicer than
![capture du 2015-01-27 10 42 33](https://cloud.githubusercontent.com/assets/930064/5916432/845dc268-a611-11e4-9d93-10d3dd7e955f.png)

Faldrian commented:
Look at the screenshots in the PR-description: There is a caret, added by the browser. It's implementation-specific if a caret is added.

I oriented the design on the desktop-layout and the select is aligned right - so this will be aligned right. It feld natural to expect it on the right when you have been using desktop version before.

Flaburgan commented:
Hm, weird you have a caret with Firefox for Android and I don't have one with Firefox desktop. Anyway, this is not critical.

About the alignment, you have my opinion, so it will be up to others to decide :)

Apart from that, it's an awesome work @Faldrian thank you so much for working on this!

marienfressinaud commented:
I also prefer the select on the left (so it is aligned with other information) but I agree with your argument to keep it on the right.

It is a minor problem anyway! ;)

Faldrian commented:
@Flaburgan Firefox for desktop uses other styles than the mobile one. It also does not have a touch-friendly overlay when selecting the dropdown (like the one shown in 2nd screenshot).

Faldrian commented:
Please correct the Readme.md when merging... when I submitted this PR, it was mergeable, but if other PRs get merged first, the state changes. I can't keep the PR up to date (rebase every time develop gets a new line in Readme.md) - so I think you can do it when merging? :)

jhass commented:
Yeah, don't worry about changelog conflicts, those should be easy enough to solve for the reviewer ;)

svbergerem commented on app/assets/javascripts/mobile/profile_aspects.js:
```diff
@@ -0,0 +1,78 @@
+
+  function profileRefreshAspectDropdown() {
+    var covertext, numselected = $('#user_aspects option.selected').length;
+    if(numselected == 0) {
```
please use `===` instead

svbergerem commented on app/assets/javascripts/mobile/profile_aspects.js:
```diff
@@ -0,0 +1,78 @@
+      covertext = t_addcontact;
+    } else {
+      $('#user_aspects').addClass('hasconnection');
+      if(numselected == 1) {
```
please use `===` instead

svbergerem commented on app/assets/stylesheets/mobile/mobile.css.scss:
```diff
@@ -1166,3 +1166,14 @@ select#aspect_ids_ {
+  padding: 3px;
+
+  &.hasconnection {
+    background-color: #C1ED94;
```
Is a color in `colors.css.scss` which you could use here? If you aren't able to find one it would be great if you could add this color to the file.

svbergerem commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+    - current_user.aspects.each do |aspect|
+      - if !contact.nil? && contact.in_aspect?(aspect)
+        - mid = contact.aspect_memberships.where(:aspect_id => aspect.id).first.id
+        %option{:value => aspect.id, 'data-name' => aspect.name, 'data-mid' => mid, :class => 'selected'}
```
please use `data-membership_id` or something similar instead of `data-mid`.

svbergerem commented on config/locales/diaspora/en.yml:
```diff
@@ -175,6 +175,7 @@ en:
     zero: "no aspects"
     one: "1 aspect"
     other: "%{count} aspects"
+    inaspects: "in {1} aspects"
```
We already have `shared.aspect_dropdown.toggle`. You are changing this with javascript so you might also want to use some appropriate translation in `javascript.en.yml` and check if we add those translations to the mobile website

Faldrian replied:
We currently have no javascript translations in the mobile website and I worked around that to have translated strings in the chooser.
I thought it was not scope of this PR to add the javascript translations and migrate the mobile interface to use them instead of some other workarounds...

I could not use the `shared.aspect_dropdown.toggle` because I had to replace it with javascript and need the raw string instead of one with already numbers in it.

svbergerem commented on config/locales/diaspora/en.yml:
```diff
@@ -825,15 +826,15 @@ en:
+
         Sign in here: %{login_url}. If you've forgotten your sign-in details, you can ask for a reminder on that page.
-        
+
```
Please keep those whitespaces. All lines belong to one email and I am not sure if this could break the translation.

Faldrian replied:
sorry, this was auto-removed by my editor. will add them again.

svbergerem commented on features/mobile/people_aspects.feature:
```diff
@@ -0,0 +1,43 @@
+      And I toggle the mobile view
+      And I am on the home page
+      And I sign in as "bob@bob.bob" on the mobile website
+        
```
trailing whitespaces

svbergerem commented on app/assets/stylesheets/colors.css.scss:
```diff
@@ -19,6 +19,7 @@ $text-dark-grey: #666666;
 $white: white;
 $black: black;
 $green: #8EDE3D;
+$lightgreen: lighten($green,20%);
```
I guess `$light-green` would be more appropriate. (you may call this nitpicking if you want to :-P )

Faldrian commented:
Added all requested changes and i18N and stuff. :)

jhass commented on app/views/people/show.mobile.haml:
```diff
@@ -10,6 +10,8 @@
         = @person.name
       %span.description
         = @person.diaspora_handle
+      .useraspects
```
Total nitpick, but most of our classes seem to use `snake_case`

Faldrian replied:
I think at first I did not know what I wanted and I started with a div with a class. I have transformed it into a div, since the select itself has some id now and I don't need this anymore.

jhass commented on app/assets/javascripts/mobile/profile_aspects.js:
```diff
@@ -0,0 +1,74 @@
+  var person_id    = $('#user_aspects').data('person-id');
+
+  function profileRefreshAspectDropdown() {
+    var covertext, numselected = $('#user_aspects option.selected').length;
```
Actually, while I'm nitpicking on `snake_case` for naming things... ;P

Faldrian replied:
numselected -> num_selected

jhass replied:
`cover_text`, `has_connection`, `list_cover`

Faldrian replied:
done

jhass commented on app/assets/javascripts/mobile/profile_aspects.js:
```diff
@@ -0,0 +1,74 @@
+      // remove from aspect
+      var membership_id = selected.data('membership_id');
+      $.ajax({
+        url: '/aspect_memberships/' + membership_id,
```
I know this is basically nowhere done yet, but let's start using [js-routes](https://github.com/railsware/js-routes) when we have it in the Gemfile already. You might need to add [the require](https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/main.js#L6) for mobile, I didn't check.

Faldrian replied:
done

jhass commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+- if user_signed_in? && @person != current_user.person
+  %select{:id => "user_aspects", 'data-person-id' => @person.id}
```
Mmh, I wonder if this has to be an ID or if we can't have this reusable already for when somebody wants to add more than one dropdown to a single page, think contacts management/overview.

Faldrian replied:
changed to .user_aspects

jhass commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+- if user_signed_in? && @person != current_user.person
```
Related to the last point, the if should probably be around the render for the partial, since it's page specific.

Faldrian replied:
done

jhass commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+      = t("add_contact")
+    - contact = current_user.contact_for(@person)
+    - current_user.aspects.each do |aspect|
+      - if !contact.nil? && contact.in_aspect?(aspect)
```
Don't `!contact.nil? &&`, just `contact &&` or even `contact.try(:in_aspect?, aspect)`

Faldrian replied:
done

jhass commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+    - contact = current_user.contact_for(@person)
+    - current_user.aspects.each do |aspect|
+      - if !contact.nil? && contact.in_aspect?(aspect)
+        - membership_id = contact.aspect_memberships.where(:aspect_id => aspect.id).first.id
```
`.limit(1).pluck(:id)` instead of `.first.id`, avoids instantiating the record just for a single attribute.

Faldrian replied:
done

jhass commented on features/mobile/people_aspects.feature:
```diff
@@ -0,0 +1,43 @@
+      And I sign in as "bob@bob.bob" on the mobile website
+
+    Scenario: verify different states of the cover button
+      When I visit the user profile "alice"
```
https://github.com/diaspora/diaspora/blob/develop/features/desktop/connects_users.feature#L11 No need for a new step

Faldrian replied:
removed

svbergerem commented on config/locales/diaspora/en.yml:
```diff
@@ -1106,6 +1106,8 @@ en:
     aspect_dropdown:
       add_to_aspect: "Add contact"
+      mobile_row_checked: "✓ %{name} (remove)"
+      mobile_row_unchecked: "– %{name} (add)"
```
@jhass What do you think about removing '✓' and '-' from those translations and adding them via css ( `option.selected::before { content: '✓'; }`) instead?

jhass replied:
Sounds like a fairly sane thing to do

Faldrian replied:
Have already checked. This works on desktop browsers, but in mobile browsers the dropdown is rendered as popup-modal-dialog on your phone and that dialog does not respect any css or anything other than "this is the utf-8 text of the option". Not even background-color or anything. So ::before won't work.

svbergerem replied:
In that case you could add the additional char in `app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml` and `app/assets/javascripts/mobile/profile_aspects.js`. I just think those chars shouldn't be in the translated string.

Faldrian replied:
will do. My thought was just "if we do string concatenation, do it all at once".

Faldrian commented:
I rewrote some of the code...

- can be used multiple times on a page (just render the template multiple times with different ids as parameter)
- this instance still has `#user_aspects` since cucumber can not test with just using a class

the code is now much cleaner and "self-contained" in a way that it does not rely on there being only this one instance. :)

jhass commented on features/step_definitions/mobile_steps.rb:
```diff
@@ -33,3 +33,7 @@
 end
+
+Then /^the aspect dropdown should be labeled "([^"]*)"/ do |label|
+  expect(page).to have_css("#user_aspects option.list_cover", :text => label)
```
This is where you couldn't use the class?

Faldrian replied:
No, there I could have used `.user_aspects`. This is where I could not use the class: https://github.com/diaspora/diaspora/pull/5594/files#diff-07ff6d8f76a7cabb01ec40c8990d498bR19

jhass replied:
That ends up here https://github.com/diaspora/diaspora/blob/develop/features/step_definitions/web_steps.rb#L75

Let's check [the docs](http://www.rubydoc.info/gems/capybara/Capybara/Node/Actions#select-instance_method), so we can select it by it's ID, name and label text. ID falls short, that needs to be unique inside the entire page, leaving us with name and label text. We have no label, so we're left with name. When we look at the step definition again, we can notice that it allows us to scope the lookup into a parent container. Seems like a reasonable alternative to make this fully reusable.

jhass replied:
Oh and btw., I wouldn't have any objections to have equivalents to https://github.com/diaspora/diaspora/blob/develop/features/step_definitions/aspects_steps.rb#L55 etc. for the mobile version of the aspect dropdown.

Faldrian replied:
used name and added "within" to have a reusable step :)

jhass commented on features/step_definitions/profile_steps.rb:
```diff
@@ -11,3 +11,6 @@
 
+When(/^I visit the user profile "([^"]*)"/) do |userhandle|
+  visit("/u/#{userhandle}")
+end
```
Didn't we want to replace this step with the existing one?

jhass commented on app/views/aspect_memberships/_aspect_membership_dropdown.mobile.haml:
```diff
@@ -0,0 +1,13 @@
+      - if contact.try(:in_aspect?, aspect)
+        - membership_id = contact.aspect_memberships.where(:aspect_id => aspect.id).limit(1).pluck(:id).first
+        %option{:value => aspect.id, 'data-name' => aspect.name, 'data-membership_id' => membership_id, :class => 'selected'}
+          = "✓ " + t('shared.aspect_dropdown.mobile_row_checked', {:name => aspect.name} )
```
Let's use string interpolation here, `"✓ #{t('shared.aspect_dropdown.mobile_row_checked', name: aspect.name)}"`

Faldrian commented:
Fixed that, thanks for the advice. :)

jhass commented:
Okay, thank you!

Flaburgan commented:
THIS IS AWESOME!

Flaburgan commented:
@Faldrian can you indicate on [bounty source](https://www.bountysource.com/issues/192899-unable-to-add-contacts-in-mobile) that you fixed the issue?

Flaburgan commented:
You can claim the bounty [here](https://www.bountysource.com/issues/192899-unable-to-add-contacts-in-mobile/developers)

goobertron commented:
Great work, Faldrian (and commenters) - thanks very much!
